x-pack/filebeat/input/httpjson/json_parser.go

```diff
@@ -0,0 +1,143 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package httpjson
+
+import (
```

Comment:
In general this file contains a very low level JSON manipulation, I'm just wondering why it is needed.

Reply:
I tried to use the Mapstr package but it is not able to parse []interface{} types of data so I wrote this code to collect those kinds of info from JSON.

Reply:
Unfortunately there is no way to parse any valid JSON as a `map[string]interface{}`. One thing you can do is to unmarshal the JSON into a empty interface (`interface{}`) and then do a type assertion/type switch to manipulate it. [I made a quick example in the Go Playground](https://go.dev/play/p/rmqk9-KHDqS) where you can see different JSONs being parsed into a empty interface and assuming different types.

I would like to better understand the need for this special JSON manipulation because the current implementation mixes JSON parsing with some trasversing/filtering. Look at the [`parse`](https://github.com/kush-elastic/beats/blob/664d95b1b043b3199e88156f72c488928bb175f3/x-pack/filebeat/input/httpjson/json_parser.go#L15) function: it receives a JSON and a key, splits the key on the `.` (so it assumes a specific JSON structure), does some data manipulation and returns a `[]string`. That is definitely not a generic JSON parser.

This makes me wonder if this whole JSON manipulation belongs to the `httpjson` package or should be implemented somewhere else.

There are also some hard to grasp constructions and many type casts that could lead to bugs/panics.

Would you like to setup a call to discuss this PR face-2-face?

Reply:
Thanks for the detailed analysis.
Let's connect at your convenience, that would be helpful.